// @ionic/app-scripts/dist/util/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FILE_CHANGE_EVENT = 'change';
exports.FILE_ADD_EVENT = 'add';
exports.FILE_DELETE_EVENT = 'unlink';
exports.DIRECTORY_ADD_EVENT = 'addDir';
exports.DIRECTORY_DELETE_EVENT = 'unlinkDir';
exports.SOURCE_MAP_TYPE_CHEAP = 'eval';
exports.SOURCE_MAP_TYPE_EXPENSIVE = 'source-map';
exports.BUILD_DIR = 'build';
exports.SRC_DIR = 'src';
exports.TMP_DIR = '.tmp';
exports.SOURCEMAP_DIR = '.sourcemaps';
exports.WWW_DIR = 'www';
exports.NODE_MODULES = 'node_modules';
exports.IONIC_ANGULAR = 'ionic-angular';
exports.COMPONENT = 'component';
exports.DIRECTIVE = 'directive';
exports.PAGE = 'page';
exports.PIPE = 'pipe';
exports.PROVIDER = 'provider';
exports.TABS = 'tabs';
exports.AT_ANGULAR = '@angular';
exports.RXJS = 'rxjs';
exports.CORDOVA = 'cordova';
exports.TYPESCRIPT = 'typescript';
exports.ENV_VAR_PROD = 'prod';
exports.ENV_VAR_DEV = 'dev';
exports.ENV_VAR_IONIC_ENV = 'IONIC_ENV';
exports.ENV_VAR_IONIC_AOT = 'IONIC_AOT';
exports.ENV_VAR_IONIC_MINIFY_JS = 'IONIC_MINIFY_JS';
exports.ENV_VAR_IONIC_MINIFY_CSS = 'IONIC_MINIFY_CSS';
exports.ENV_VAR_IONIC_OPTIMIZE_JS = 'IONIC_OPTIMIZE_JS';
exports.ENV_VAR_ROOT_DIR = 'IONIC_ROOT_DIR';
exports.ENV_VAR_SRC_DIR = 'IONIC_SRC_DIR';
exports.ENV_VAR_DEEPLINKS_DIR = 'IONIC_DEEPLINKS_DIR';
exports.ENV_VAR_PAGES_DIR = 'IONIC_PAGES_DIR';
exports.ENV_VAR_COMPONENTS_DIR = 'IONIC_COMPONENTS_DIR';
exports.ENV_VAR_DIRECTIVES_DIR = 'IONIC_DIRECTIVES_DIR';
exports.ENV_VAR_PIPES_DIR = 'IONIC_PIPES_DIR';
exports.ENV_VAR_PROVIDERS_DIR = 'IONIC_PROVIDERS_DIR';
exports.ENV_VAR_TMP_DIR = 'IONIC_TMP_DIR';
exports.ENV_VAR_WWW_DIR = 'IONIC_WWW_DIR';
exports.ENV_VAR_FONTS_DIR = 'IONIC_FONTS_DIR';
exports.ENV_VAR_SOURCEMAP_DIR = 'IONIC_SOURCEMAP_DIR';
exports.ENV_VAR_HTML_TO_SERVE = 'IONIC_HTML_TO_SERVE';
exports.ENV_VAR_BUILD_DIR = 'IONIC_BUILD_DIR';
exports.ENV_VAR_NODE_MODULES_DIR = 'IONIC_NODE_MODULES_DIR';
exports.ENV_VAR_ANGULAR_CORE_DIR = 'IONIC_ANGULAR_CORE_DIR';
exports.ENV_VAR_TYPESCRIPT_DIR = 'IONIC_TYPESCRIPT_DIR';
exports.ENV_VAR_RXJS_DIR = 'IONIC_RXJS_DIR';
exports.ENV_VAR_IONIC_ANGULAR_DIR = 'IONIC_ANGULAR_DIR';
exports.ENV_VAR_CORE_COMPILER_FILE_PATH = 'IONIC_CORE_COMPILER_FILE_PATH';
exports.ENV_VAR_CORE_DIR = 'IONIC_CORE_DIR';
exports.ENV_VAR_IONIC_ANGULAR_TEMPLATE_DIR = 'IONIC_ANGULAR_TEMPLATE_DIR';
exports.ENV_VAR_TARGET = 'IONIC_TARGET';
exports.ENV_VAR_PLATFORM = 'IONIC_PLATFORM';
exports.ENV_VAR_IONIC_ANGULAR_ENTRY_POINT = 'IONIC_ANGULAR_ENTRY_POINT';
exports.ENV_VAR_APP_SCRIPTS_DIR = 'IONIC_APP_SCRIPTS_DIR';
exports.ENV_VAR_GENERATE_SOURCE_MAP = 'IONIC_GENERATE_SOURCE_MAP';
exports.ENV_VAR_SOURCE_MAP_TYPE = 'IONIC_SOURCE_MAP_TYPE';
exports.ENV_VAR_MOVE_SOURCE_MAPS = 'IONIC_MOVE_SOURCE_MAPS';
exports.ENV_TS_CONFIG = 'IONIC_TS_CONFIG';
exports.ENV_APP_ENTRY_POINT = 'IONIC_APP_ENTRY_POINT';
exports.ENV_APP_NG_MODULE_PATH = 'IONIC_APP_NG_MODULE_PATH';
exports.ENV_APP_NG_MODULE_CLASS = 'IONIC_APP_NG_MODULE_CLASS';
exports.ENV_COMPONENTS_NG_MODULE_PATH = 'IONIC_COMPONENTS_NG_MODULE_PATH';
exports.ENV_PIPES_NG_MODULE_PATH = 'IONIC_PIPES_NG_MODULE_PATH';
exports.ENV_DIRECTIVES_NG_MODULE_PATH = 'IONIC_DIRECTIVES_NG_MODULE_PATH';
exports.ENV_GLOB_UTIL = 'IONIC_GLOB_UTIL';
exports.ENV_CLEAN_BEFORE_COPY = 'IONIC_CLEAN_BEFORE_COPY';
exports.ENV_READ_CONFIG_JSON = 'IONIC_READ_CONFIG_JSON';
exports.ENV_OUTPUT_JS_FILE_NAME = 'IONIC_OUTPUT_JS_FILE_NAME';
exports.ENV_OUTPUT_CSS_FILE_NAME = 'IONIC_OUTPUT_CSS_FILE_NAME';
exports.ENV_WEBPACK_FACTORY = 'IONIC_WEBPACK_FACTORY';
exports.ENV_WEBPACK_LOADER = 'IONIC_WEBPACK_LOADER';
exports.ENV_CACHE_LOADER = 'IONIC_CACHE_LOADER';
exports.ENV_AOT_WRITE_TO_DISK = 'IONIC_AOT_WRITE_TO_DISK';
exports.ENV_BAIL_ON_LINT_ERROR = 'IONIC_BAIL_ON_LINT_ERROR';
exports.ENV_TYPE_CHECK_ON_LINT = 'IONIC_TYPE_CHECK_ON_LINT';
exports.ENV_ENABLE_LINT = 'IONIC_ENABLE_LINT';
exports.ENV_DISABLE_LOGGING = 'IONIC_DISABLE_LOGGING';
exports.ENV_START_WATCH_TIMEOUT = 'IONIC_START_WATCH_TIMEOUT';
exports.ENV_NG_MODULE_FILE_NAME_SUFFIX = 'IONIC_NG_MODULE_FILENAME_SUFFIX';
exports.ENV_POLYFILL_FILE_NAME = 'IONIC_POLYFILL_FILE_NAME';
exports.ENV_PRINT_WEBPACK_DEPENDENCY_TREE = 'IONIC_PRINT_WEBPACK_DEPENDENCY_TREE';
exports.ENV_PARSE_DEEPLINKS = 'IONIC_PARSE_DEEPLINKS';
exports.ENV_PURGE_UNUSED_FONTS = 'IONIC_PURGE_UNUSED_FONTS';
exports.ENV_SKIP_IONIC_ANGULAR_VERSION = 'IONIC_SKIP_IONIC_ANGULAR_VERSION';
/* Providers */
exports.ENV_ACTION_SHEET_CONTROLLER_CLASSNAME = 'IONIC_ACTION_SHEET_CONTROLLER_CLASSNAME';
exports.ENV_ACTION_SHEET_CONTROLLER_PATH = 'IONIC_ACTION_SHEET_CONTROLLER_PATH';
exports.ENV_ACTION_SHEET_VIEW_CONTROLLER_PATH = 'IONIC_ACTION_SHEET_VIEW_CONTROLLER_PATH';
exports.ENV_ACTION_SHEET_COMPONENT_PATH = 'IONIC_ACTION_SHEET_COMPONENT_PATH';
exports.ENV_ACTION_SHEET_COMPONENT_FACTORY_PATH = 'IONIC_ACTION_SHEET_COMPONENT_FACTORY_PATH';
exports.ENV_ALERT_CONTROLLER_CLASSNAME = 'IONIC_ALERT_CONTROLLER_CLASSNAME';
exports.ENV_ALERT_CONTROLLER_PATH = 'IONIC_ALERT_CONTROLLER_PATH';
exports.ENV_ALERT_VIEW_CONTROLLER_PATH = 'IONIC_ALERT_VIEW_CONTROLLER_PATH';
exports.ENV_ALERT_COMPONENT_PATH = 'IONIC_ALERT_COMPONENT_PATH';
exports.ENV_ALERT_COMPONENT_FACTORY_PATH = 'IONIC_ALERT_COMPONENT_FACTORY_PATH';
exports.ENV_APP_ROOT_COMPONENT_PATH = 'IONIC_APP_ROOT_COMPONENT_PATH';
exports.ENV_LOADING_CONTROLLER_CLASSNAME = 'IONIC_LOADING_CONTROLLER_CLASSNAME';
exports.ENV_LOADING_CONTROLLER_PATH = 'IONIC_LOADING_CONTROLLER_PATH';
exports.ENV_LOADING_VIEW_CONTROLLER_PATH = 'IONIC_LOADING_VIEW_CONTROLLER_PATH';
exports.ENV_LOADING_COMPONENT_PATH = 'IONIC_LOADING_COMPONENT_PATH';
exports.ENV_LOADING_COMPONENT_FACTORY_PATH = 'IONIC_LOADING_COMPONENT_FACTORY_PATH';
exports.ENV_MODAL_CONTROLLER_CLASSNAME = 'IONIC_MODAL_CONTROLLER_CLASSNAME';
exports.ENV_MODAL_CONTROLLER_PATH = 'IONIC_MODAL_CONTROLLER_PATH';
exports.ENV_MODAL_VIEW_CONTROLLER_PATH = 'IONIC_MODAL_VIEW_CONTROLLER_PATH';
exports.ENV_MODAL_COMPONENT_PATH = 'IONIC_MODAL_COMPONENT_PATH';
exports.ENV_MODAL_COMPONENT_FACTORY_PATH = 'IONIC_MODAL_COMPONENT_FACTORY_PATH';
exports.ENV_PICKER_CONTROLLER_CLASSNAME = 'IONIC_PICKER_CONTROLLER_CLASSNAME';
exports.ENV_PICKER_CONTROLLER_PATH = 'IONIC_PICKER_CONTROLLER_PATH';
exports.ENV_PICKER_VIEW_CONTROLLER_PATH = 'IONIC_PICKER_VIEW_CONTROLLER_PATH';
exports.ENV_PICKER_COMPONENT_PATH = 'IONIC_PICKER_COMPONENT_PATH';
exports.ENV_PICKER_COMPONENT_FACTORY_PATH = 'IONIC_PICKER_COMPONENT_FACTORY_PATH';
exports.ENV_POPOVER_CONTROLLER_CLASSNAME = 'IONIC_POPOVER_CONTROLLER_CLASSNAME';
exports.ENV_POPOVER_CONTROLLER_PATH = 'IONIC_POPOVER_CONTROLLER_PATH';
exports.ENV_POPOVER_VIEW_CONTROLLER_PATH = 'IONIC_POPOVER_VIEW_CONTROLLER_PATH';
exports.ENV_POPOVER_COMPONENT_PATH = 'IONIC_POPOVER_COMPONENT_PATH';
exports.ENV_POPOVER_COMPONENT_FACTORY_PATH = 'IONIC_POPOVER_COMPONENT_FACTORY_PATH';
exports.ENV_SELECT_POPOVER_CLASSNAME = 'IONIC_SELECT_POPOVER_CLASSNAME';
exports.ENV_SELECT_POPOVER_COMPONENT_PATH = 'IONIC_SELECT_POPOVER_COMPONENT_PATH';
exports.ENV_SELECT_POPOVER_COMPONENT_FACTORY_PATH = 'IONIC_SELECT_POPOVER_COMPONENT_FACTORY_PATH';
exports.ENV_TOAST_CONTROLLER_CLASSNAME = 'IONIC_TOAST_CONTROLLER_CLASSNAME';
exports.ENV_TOAST_CONTROLLER_PATH = 'IONIC_TOAST_CONTROLLER_PATH';
exports.ENV_TOAST_VIEW_CONTROLLER_PATH = 'IONIC_TOAST_VIEW_CONTROLLER_PATH';
exports.ENV_TOAST_COMPONENT_PATH = 'IONIC_TOAST_COMPONENT_PATH';
exports.ENV_TOAST_COMPONENT_FACTORY_PATH = 'IONIC_TOAST_COMPONENT_FACTORY_PATH';
exports.BUNDLER_WEBPACK = 'webpack';
